codalab/server/bundle_manager.py

```diff
@@ -388,6 +388,15 @@ def _filter_and_sort_workers(self, workers_list, bundle, bundle_resources):
         # Filter by tag.
         if bundle.metadata.request_queue:
             workers_list = self._get_matched_workers(bundle.metadata.request_queue, workers_list)
+        else:
+            # The bundle is untagged, so we want to keep workers that are not
+            # tag-exclusive or don't have a tag defined.
+            # (removing workers that are tag_exclusive and have tags defined).
+            workers_list = [
+                worker
+                for worker in workers_list
+                if not worker['tag_exclusive'] or not worker['tag']
```

Comment:
I just want to understand the relationship between `tag` and `tag_exclusive`? My understanding is that there could be 4 possibilities as follows:

| | tag |  tag_exclusive |
| -- | ------------- | ----- |
|1 | null | True |
|2| tagged | True |
|3| null | False |
|4| tagged | False | 

`tag_exclusive` will take effect in the 2nd case, right?

Reply:
Under the current implementation of tagging, yes. Note that the logic still applies in case (1)---the worker would take only untagged bundles. But this is the current behavior, anyway.

Reply:
I guess I was thinking the case that a user has 2 private workers attached to the server. Both of them have their tag set. The 1st worker sets the `tag_exclusive` to be `True`, whereas the 2nd sets the `tag_exclusive` to be `False`. When this user submits a bundle without a `request-queue`, do we want to route this bundle to run on the 2nd worker? If we do, we might want to remove the logic `not worker['tag']`...

Reply:
Wouldn't the current logic route the bundle to the second worker?

```
In [1]: worker = {"tag_exclusive": False, "tag": "sometag"}

In [2]: not worker['tag_exclusive'] or not worker['tag']
Out[2]: True
```

Reply:
Yes. That's true. Never mind~ We still want to keep those untagged workers around in case the bundle aren't able to schedule to run on the 2nd private worker, it can still be routed to those untagged workers.